zuul-core/src/main/java/com/netflix/zuul/message/http/HttpRequestMessageImpl.java

```diff
@@ -510,12 +514,23 @@ protected String generateInfoForLogging() {
     @Override
     public String getOriginalHost() {
         try {
-            return getOriginalHost(getHeaders(), getServerName());
+            if (!CACHE_ORIGINAL_HOST.get()) {
+                return generateOriginalHost(getHeaders(), getServerName());
+            }
+
+            if (originalHost == null) {
+                originalHost = generateOriginalHost(getHeaders(), getServerName());
+            }
+            return originalHost;
         } catch (URISyntaxException e) {
             throw new IllegalArgumentException(e);
         }
     }
 
+    String generateOriginalHost(Headers headers, String serverName) throws URISyntaxException {
```

Comment:
Think you can remove this method and just call `getOriginalHost(headers, serverName)` from within the `getOriginalHost()` method